File: internal/utils/subdomain.go
package utils

import (
	"fmt"
	"regexp"
	"strings"
)

var subdomainPattern = regexp.MustCompile(`^[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?$`)

func NormalizeSubdomain(subdomain string) string {
	return strings.ToLower(strings.TrimSpace(subdomain))
}

func ValidateSubdomain(subdomain string) error {
	if !subdomainPattern.MatchString(subdomain) {
		return fmt.Errorf("invalid subdomain %q: use 1-63 lowercase letters, numbers, or internal hyphens", subdomain)
	}

	return nil
}

File: internal/server/admin/web-v2/src/lib/subdomain.ts
const subdomainPattern = /^[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?$/

export function normalizeSubdomain(value: string): string {
  return value.trim().toLowerCase()
}

export function validateSubdomain(value: string): string | null {
  const normalized = normalizeSubdomain(value)
  if (!normalized) return "Enter a subdomain"
  if (!subdomainPattern.test(normalized)) {
    return "Use 1–63 letters, numbers, or internal hyphens"
  }
  return null
}
